go/saltpack/encrypt.go

```diff
@@ -176,16 +168,41 @@ func (es *encryptStream) init(sender BoxSecretKey, receivers []BoxPublicKey) err
 		}
 
 		eh.Receivers = append(eh.Receivers, keys)
+	}
 
-		tagKey := ephemeralShared
-		if !senderAnon {
-			tagKey = sender.Precompute(receiver)
-		}
-		es.tagKeys = append(es.tagKeys, tagKey)
+	// Encode the header and the header length, and write them out immediately.
+	headerBytes, err := encodeToBytes(es.header)
+	if err != nil {
+		return err
+	}
+	headerLen, err := encodeToBytes(len(headerBytes))
+	if err != nil {
+		return err
+	}
+	headerHash := sha512.Sum512(headerBytes)
+	es.headerHash = headerHash[:]
+	_, err = es.output.Write(headerLen)
+	if err != nil {
+		return err
 	}
+	_, err = es.output.Write(headerBytes)
+	if err != nil {
+		return err
+	}
+
+	// Use the header hash to compute the MAC keys.
+	es.computeMACKeys(sender, receivers)
+
 	return nil
 }
 
+func (es *encryptStream) computeMACKeys(sender BoxSecretKey, receivers []BoxPublicKey) {
+	for _, receiver := range receivers {
+		macKeyBox := sender.Box(receiver, nonceForMACKeyBox(es.headerHash[:]), make([]byte, 32))
```

Comment:
I ended up factoring this out into a separate function, `computeMACKey`. That means that I've failed to reuse the 32-byte zero buffer, but I think it's nicer to have all the details hidden. Let me know if you feel differently.